### ISSUE TITLE
Add a string endpoint description plugin for twisted.internet.endpoints.TCP6ServerEndpoint

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5699]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)| @ashfall reported|
|-|-|
|Trac ID|trac#5699|
|Type|enhancement|
|Created|2012-06-01 10:13:12Z|
|Branch|https://github.com/twisted/twisted/tree/tcp6-server-string-description-plugin-5699|

A string endpoint description plugin for this TCP server endpoint that supports IPv6 addresses (added in [#5694](https://github.com/twisted/twisted/issues/5694)) is needed so that the endpoint is accessible from the command-line.

<details><summary>Searchable metadata</summary>

```
trac-id__5699 5699
type__enhancement enhancement
reporter__ashfall ashfall
priority__normal normal
milestone__ 
branch__branches_tcp6_server_string_description_plugin_5699 branches/tcp6-server-string-description-plugin-5699
branch_author__ashfall ashfall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1338545592000000 1338545592000000
changetime__1341402413000000 1341402413000000
version__None None
owner__ashfall ashfall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)|@ashfall set owner to @ashfall|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_2"></a>|@ashfall commented|
|-|-|

(In [34609]) Branching to 'tcp6-server-string-description-plugin-5699'

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_3"></a>|@ashfall removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tcp6-server-string-description-plugin-5699)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks!  This looks pretty good.  A couple things are missing though, I think:

  1. There aren't any tests that actually exercise the new parser.  See, eg, `_parseStreamServerTest` (combined with, eg, `test_parseStreamServerINET`) in `SystemdEndpointPluginTests`.  In other words, a test should call `parseStreamServer` with a "tcp6:" endpoint string description.
  1. Hard-coding the backlog to 50 doesn't seem necessary.  I don't think the backlog argument is actually very important (almost no one ever wants to specify a value for it), but I wonder why you didn't allow it to be supplied in the string.  "tcp:1234:backlog=50" works for TCP4 servers, so it's surprising for it to not work for TCP6 servers, I think.  On the other hand, if we want to make an explicit, intentional decision to not support it in TCP6, and drop support for it in TCP4 eventually, that may be fine too.  So, mostly I wonder why it's hard-coded instead of parameterized.  (Also, it is documented as a parameter in `_parseServer`.)
  1. The new class is private, and generally we don't need to bother to advertise private classes in news fragments.  More importantly is the end-user facing feature, which is that IPv6 TCP servers can now be created using the endpoint string description feature.  That's what the news fragment should talk about (perhaps mentioning `twisted.internet.endpoints.serverFromString`).
Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_5"></a>|@ashfall removed owner|
|-|-|

Replying to [exarkun](#note_4):
 
>   1. There aren't any tests that actually exercise the new parser.  See, eg, `_parseStreamServerTest` (combined with, eg, `test_parseStreamServerINET`) in `SystemdEndpointPluginTests`.  In other words, a test should call `parseStreamServer` with a "tcp6:" endpoint string description.

Added a test that uses `_parseStreamServerTest` in `TCP6ServerEndpointPluginTests`, I think that should suffice.

Fixed the other issues as well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks.  There's still no good test coverage for the actual format for this endpoint plugin parser.  "8080" is not a TCP6 server endpoint string!  It's a TCP4 server endpoint string.  `test_parseStreamServerPort` cheats by instantiating `endpoints._TCP6ServerParser` instead of calling `serverFromString`.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_7"></a>|@ashfall removed owner|
|-|-|

Added the said test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks!  Looks great now.  Please merge. :)

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_9"></a>|@ashfall set status to `closed`|
|-|-|

(In [34680]) Merge tcp6-server-string-description-plugin-5699: A string description plugin for the TCP6ServerEndpoint.

Author: ashfall
Reviewer: exarkun
Fixes: [#5699](https://github.com/twisted/twisted/issues/5699)

Added a string endpoint description plugin for the TCP6ServerEndpoint so that it can be accessed from serverFromString.
